### ISSUE TITLE
AndroidInjection doesn't allow to inject a super class

### ISSUE THREAD

**Issue by LuigiPapino** [google/dagger#653]
I discovered this issue playing with AndroidAnnotations but it should be the same for every scenario where a Fragment extends another Fragment and the base one use `AndroindInjection.inject`.

The detailed scenario: 

MyFragment use AndroidInjection to inject the presenter. Following the Subcomponent, module and Fragment:
```
@Subcomponent
public interface MyFragmentSubComponent extends AndroidInjector<MyFragment> {

  @Subcomponent.Builder
  abstract class Builder extends AndroidInjector.Builder<MyFragment> {
  }
}
```

```
@Module(subcomponents = { MyFragmentSubComponent.class })
abstract class FragmentModule {

  @Binds
  @IntoMap
  @FragmentKey(MyFragment.class)
  abstract AndroidInjector.Factory<? extends Fragment> bindMyFragmentInjectorFactory(
      MyFragmentComponent.Builder builder);
}
```

```
@EFragment(R.layout.screen_my_fragment)
public class RMyFragment extends BaseFragment implements RecipeBrowserPageView {
 @Inject
  MyPresenter presenter;

  @Override
  public void onAttach(Context context) {
    AndroidSupportInjection.inject(this);
    super.onAttach(context);
  }
}
```

``

AndroidAnnotations will generate the class MyFragment_ extending MyFragment.
If I try to use MyFragment_ this error will be raised: 
`                                                                                 Caused by: java.lang.IllegalArgumentException: No injector factory bound for Class<com.package.MyFragment_>. Injector factories were bound for supertypes of com.package.MyFragment_: [com.package.MyFragment]. Did you mean to bind an injector factory for the subtype?
` 

Because obviously the subcomponent is for the base MyFragment. 

I'd like to try adding a class parameter to AndroidInjector.inject, and force to look for the correct injector based on the class parameter, and inject the instance parameter. 
It seems possible to you?

Could anyone suggest a different approach or a solution for this problem?

Thanks very much.

**Comment by UsherBaby** [comment]
@LuigiPapino Add your FragmentModule to parent component

**Comment by LuigiPapino** [comment]
@UsherBaby thanks for the advice. The project is setup correctly, with the module added to the parent component. I didn't include in the example for shortness.

In other words, the problem is that AndroidInjector.inject look for the class name at runtime, so it's not able to inject a super class.

Here the relevant piece of code: 
```
  /**
   * Attempts to perform members-injection on {@code instance}, returning {@code true} if
   * successful, {@code false} otherwise.
   *
   * @throws InvalidInjectorBindingException if the injector factory bound for a class does not
   *     inject instances of that class
   */
  public boolean maybeInject(T instance) {
    Provider<AndroidInjector.Factory<? extends T>> factoryProvider =
        injectorFactories.get(instance.getClass());
    if (factoryProvider == null) {
      return false;
    }

    @SuppressWarnings("unchecked")
    AndroidInjector.Factory<T> factory = (AndroidInjector.Factory<T>) factoryProvider.get();
    try {
      AndroidInjector<T> injector =
          checkNotNull(
              factory.create(instance),
              "%s.create(I) should not return null.",
              factory.getClass().getCanonicalName());

      injector.inject(instance);
      return true;
    } catch (ClassCastException e) {
      throw new InvalidInjectorBindingException(
          String.format(
              "%s does not implement AndroidInjector.Factory<%s>",
              factory.getClass().getCanonicalName(), instance.getClass().getCanonicalName()),
          e);
    }
  }
```

However I'll prepare a project as reference to explain better the scenario.

**Comment by LuigiPapino** [comment]
Here the project that replicate the scenario: https://github.com/LuigiPapino/DaggerAndroidInjectorSuperClass/tree/master

A `BaseActivity` with a `BasePresenter` injected through the `AndriondInection.inject`;
Then a `MainActivity` extends the `BaseActivity`.

Launching the MainActivity will raise this exception:
`Caused by: java.lang.IllegalArgumentException: No injector factory bound for Class<net.dragora.daggerandroidinjectorsuperclass.MainActivity>. Injector factories were bound for supertypes of net.dragora.daggerandroidinjectorsuperclass.MainActivity: [net.dragora.daggerandroidinjectorsuperclass.BaseActivity]. Did you mean to bind an injector factory for the subtype?`

**Comment by ronshapiro** [comment]
You need to have the `AndroidInjector` bound for the actual type that's being injected. Members injection can have weird implications for subtypes that have injected members of their own. In your initial code, switch to `AndroidInjector<MyFragment_>` and `@ActivityKey(MyFragment_)`

**Comment by LuigiPapino** [comment]
This has been the first solution came in my mind, and I received a build error that I don't remember now.

I tried again now, and works well. Don't know what mess I did. 
Thanks very much!

**Comment by skhaz** [comment]
Hi @LuigiPapino you said that your demo works well. I cloned your repository and had the same issue, do you mind to check if some changes was not pushed to your demo?

**Comment by ohadnav** [comment]
Could you upload the solution to your demo app? thanks! :)
@ronshapiro could you be more specific as which line to change, and how to change it (exactly)?

**Comment by mcmorning** [comment]
I had a same issue, too. To fix this issue, every subclass has to be added in the module explicitly even if subclass doesn't have any @Inject.

For example, if there are BaseFragment and MainFragment..

@Module
public abstract class FragmentModule {
    @ContributesAndroidInjector
    abstract BaseFragment contributeBaseFragmentInjector();

    @ContributesAndroidInjector
    abstract MainFragment contributeMainFragmentInjector();
}

And if you wanna fix @LuigiPapino's code, change [ActivityModule.java:L19](https://github.com/LuigiPapino/DaggerAndroidInjectorSuperClass/blob/master/app/src/main/java/net/dragora/daggerandroidinjectorsuperclass/di/ActivityModule.java#L19), BaseActivity.class -> MainActivity.class

**Comment by azatmar** [comment]
Is it possible to have the members injected for the sub classes too without specifying the sub-class name? This would be really nice, where you could use the injection interfaces among different libraries.

**Comment by Belka1000867** [comment]
Have the same problem.
Is there any solution so far?

**Comment by maher640** [comment]
I've came across the same struggle, but now the case is that the BaseActivity is injected while many sub-activities weren't injected or needed to be. I was left out with either adding them to Dagger or something else. I came up with this solution, not the best on the planet by saved my ass from creating many useless Dagger classes.

![untitled drawing 1](https://user-images.githubusercontent.com/4786082/50382373-ca52a200-06ae-11e9-9213-9998947a286d.jpg)


Now if an activity will be injected, it'll inherit the Base with injection and if it's not it'll inherit the Base without injection.
